flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java

```diff
@@ -642,7 +643,31 @@ private static class WatermarkAggregator<T> {
          *     Optional.empty()} otherwise.
          */
         public Optional<Watermark> aggregate(T key, Watermark watermark) {
-            watermarks.put(key, watermark);
+            Watermark oldWatermark = watermarks.put(key, watermark);
+            // Step (1): Update the latest watermark of current key as the aggregatedWatermark
+            // directly if it is less than the aggregatedWatermark.
+            if (watermark.getTimestamp() < aggregatedWatermark.getTimestamp()) {
+                aggregatedWatermark = watermark;
+                return Optional.of(aggregatedWatermark);
+            }
+
+            // Step(2): The aggWM won't change when these conditions are met, so return directly:
+            // case1. The latest WM of the current key isn't changed
+            // case2. When oldWatermark isn't null and is greater than aggWm, it means that aggWm
+            // comes from other keys. If new WM is greater than or equal to aggWm, then aggWm must
+            // not change.
+            // case3. When oldWatermark is null and {@link watermarks} has other keys, it means that
+            // aggWm comes from other keys. If new WM is greater than or equal to aggWm, then aggWm
+            // must not change.
+            // Note: Here's an implicit condition `watermark >= aggregatedWatermark` due to step(1),
+            // that's why it's not written in the code of case2 and case3.
+            if (watermark.equals(oldWatermark)
+                    || (oldWatermark != null
+                            && oldWatermark.getTimestamp() > aggregatedWatermark.getTimestamp())
+                    || (oldWatermark == null && watermarks.size() > 1)) {
```

Comment:
nit
```suggestion
                    || (oldWatermark == null && !watermarks.isEmpty())) {
```

Reply:
`!watermarks.isEmpty()` is similar to `watermarks.size() > 0`, here is `> 1`.

Because, `watermarks.put(key, watermark)` is called first.